### ISSUE TITLE
Errors from dependent effects are propagated to value setter

### ISSUE THREAD

**Issue by felixschorer** [preactjs/signals#241]
Given a signal `foo` and an effect which throws an error whenever `foo` holds the value `true`.  

When the effect is created when `foo` already holds the value `true`,  
then an error will be thrown upon creating that effect.
```ts
const foo = signal(true);

// throws
effect(() => {
  if (foo.value) throw new Error("test");
});
```

However, when the effect is created while `foo` holds the value `false`,  
then an error will be thrown upon setting the value of `foo` to `true`.
```ts
const foo = signal(false);

effect(() => {
  if (foo.value) throw new Error("test");
});

// throws
foo.value = true;
```
This behavior is unexpected.
The presence of a faulty effect may break code, which otherwise works fine in isolation.
The errors caused by an effect can only appropriately be handled by the effect itself or logged.

A fix could be to re-throw those errors inside a new micro task.
- Effects could still run eagerly.
- Errors can be handled by a global error handler to simply log them.
- The code, which is setting the value, wouldn't break depending if there is a faulty effect present or not.

----

Another (probably related) issue is that effects aren't (and can't be) disposed properly, when the initial execution throws an error.

```ts
const foo = signal(true);

try {
  const dispose = effect(() => {
    if (foo.value) {
      throw new Error("test");
    } else {
      console.log(foo.value);
    }
  });
} catch (err) { /* ignore */ }

// Logs `true` to the console, i.e., the effect is still subscribed to `foo`.
// The effect also cannot be disposed as `dispose` is out of scope.
foo.value = false;
```

**Comment by marvinhagemeister** [comment]
> This behavior is unexpected.

I'd love to dive deeper on this on how you reached that conclusion. In both examples provided I'd expect an error to be thrown.

The last example with being unable to dispose them seems like a bug that we should address regardless.

**Comment by developit** [comment]
Might make sense to mark this issue as fixed? #245 addresses the idiosyncrasy between initial and update error propagation, which is the "unexpected" part.

Direct error propagation to the setter is an explicit design choice we made in order to avoid Signals inventing a new error propagation/handling mechanism where one already exists. The error propagation/handling logic in Signals isn't supposed to allow applications to continue to function partially in an error state, it's to prevent the graph from being incorrect such that applications with error recovery logic can function in a recovered state.

**Comment by marvinhagemeister** [comment]
Good point, marking as resolved by #245

**Comment by felixschorer** [comment]
> I'd love to dive deeper on this on how you reached that conclusion. In both examples provided I'd expect an error to be thrown.

My original example contained a faulty computed signal, but I generalized it to effects. Maybe effects didn't convey why I think this is unexpected.

So, consider a computed signal `bar` which throws an error for some reason. As long as there are no subscribers on that computed signal, I can update the value of the `foo` signal without an error being thrown. Errors are propagated on pull, i.e., when accessing `bar.value`.
```js
const foo = signal(false);

const bar = computed(() => {
  if (foo.value) throw new Error("faulty effect");
  return foo.value
});

// doesn't throw
foo.value = true
```

However, as soon as something subscribes to that faulty computed signal, errors are propagated on pull and on push. I would expect errors to flow in only one direction.
```js
const foo = signal(false);

const bar = computed(() => {
  if (foo.value) throw new Error("faulty effect");
  return foo.value
});

effect(() => {
  // throws and propagates error to setter
  console.log(bar.value)
})

// throws
foo.value = true
```

> The error propagation/handling logic in Signals isn't supposed to allow applications to continue to function partially in an error state, it's to prevent the graph from being incorrect such that applications with error recovery logic can function in a recovered state.

Consider this simple application where the `fruit` signal is hooked up to a text input.
When the user presses a button, the `addFruit` event handler will be called which appends the `fruit` to the array of fruits and clears the input by resetting `fruit` to the empty string.

```js
const fruit = signal("")
const fruits = signal(["apple", "orange"]);

const addFruit = () => {
  if (fruit.value.length > 0 && !fruits.value.contains(fruit.value)) {
    fruits.value = [...fruits.value, fruit.value]
    fruit.value = ""
  }
}
```

In the presence of a faulty effect, which throws whenever `fruits` is changed, the input won't be cleared anymore and leaves the application in an inconsistent state. **The `fruits` signal is still updated and will contain the new fruit.**

```js
effect(() => {
  if (fruits.value.length > 2) {
    throw new Error("faulty effect")
  }
})
```

`addFruit` may be unaware that this effect exists. How should it handle that error? I argue that it can't.
